src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +378,125 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
                 segments.Add(writer.ToSourceText());
             }
         }
+
+        /// <summary>
+        /// Delegates to SourceTexts within the CompositeText to determine line information.
+        /// </summary>
+        internal sealed class CompositeLineInfo : TextLineCollection
+        {
+            private readonly CompositeText _compositeText;
+            private readonly int[] _segmentLineIndexes;
+            private readonly int _lineCount;
+
+            public CompositeLineInfo(CompositeText compositeText)
+            {
+                _compositeText = compositeText;
+                _segmentLineIndexes = new int[compositeText.Segments.Length];
+
+                for (int i = 0; i < compositeText.Segments.Length; i++)
+                {
+                    _segmentLineIndexes[i] = _lineCount;
```

Comment:
can you doc this, i'm just not getting the logic here at all :D    like why is an index being set teh line count we have?